ICRC-122: ledger blocks recording management actions -- authorised mint, authorised burn blocks (#125)
dfinity/ICRC: ICRCs/ICRC-122/ICRC-122.md

> # ICRC-122: Token Burning & Minting (Consolidated)

Reviewer: why is "Consolidated" here explicitly mentioned? can you provide me some context?

I am also wondering if ICRC-122 is meant to be an extension to ICRC-1. if it is not an extension:
- CAN it be implemented among with ICRC-1?
  - is this what we plan to do in the reference implementation?
- MUST it be implemented independent of ICRC-1?

Author: remnant of a merge

> # ICRC-122: Token Burning & Minting (Consolidated)

Author: ```suggestion
# ICRC-122: Token Burning & Minting
```

| `amount`     | `Nat`                                                        | Yes      | The number of tokens to burn. This value **MUST be greater than 0**. |
| `reason`     | `Text`                                                       | Optional | Human-readable reason for the burn. |

#### For `122mint`

| Field        | Type (ICRC-3 `Value`)                                        | Required | Description |
|--------------|--------------------------------------------------------------|----------|-------------|
| `caller`     | `Value` (Must be `variant { Blob = <principal_bytes> }`)     | Yes      | The principal that invoked the ledger method (e.g., `icrc122_mint`) causing this block to be created. |
| `to`         | `Value` (Must be `variant { Array = vec { V1 [, V2] } }`)¹ | Yes      | The account to which tokens are minted. |
| `amount`     | `Nat`                                                        | Yes      | The number of tokens to mint. This value **MUST be greater than 0**. |
| `reason`     | `Text`                                                       | Optional | Human-readable reason for the mint. |

> ¹ Where `V1` is `variant { Blob = <owner_principal> }` and `V2` is `variant { Blob = <subaccount> }`. If no subaccount exists, the `Array` contains only `V1`.

Reviewer: this seems to be duplicate information as already described in the "Common Elements" section. I think we can keep it, but I just want to make sure if it was a conscious decision to provide that information again here.

Author: Yes, it was intended like this. The first is for whoever reads the entire thing, the second is for whoever wants to immediately see the block types

## Usage

> ### Comparison with ICRC-1 Mint/Burn

Reviewer: or do we expect ICRC-122 implementers to omit the `1mint` and `1burn` block types completely and replace them with `122mint` and `122burn` ?

> ### Comparison with ICRC-1 Mint/Burn

Author: Both types of blocks (i.e. from ICRC-1 and ICRC-122) may appear in a ledger. I'll add a note in the following section noting that this is possible. I wouldn't expect it that one necessarily replaces the other.  This may be possible, e.g. by not defining a minting account (so 1mint and 1burn couldn't appear). But then, it wouldn't be possible for a party to simply burn tokens

> ### Comparison with ICRC-1 Mint/Burn

Reviewer: not defining a minting account also seems to have other implications 🤔  e.g. burned fees should be sent to the minting account according to ICRC-1.

this means that burning fees would not be possible when not defining a minting account, right?

what is the zero account that could be set instead which isn't under control of anybody? probably this should be the recommendation to avoid the scenario mentioned above.

> ### Comparison with ICRC-1 Mint/Burn

Author: Yes, not defining a minting account means that parties cannot arbitrarily burn tokens.  Sending to some prescribed address may work for all practical purposes but it doesn't have the same semantics (since the supply is not changed)

> The `122mint` and `122burn` blocks introduced in ICRC-122 serve as a standardized method for tracking token supply changes, offering more structure than the implicit mint/burn events inferred from ICRC-1 transfer blocks involving the zero account.
> 
> - **ICRC-1**: In the original ICRC-1 standard, mint and burn events were typically represented by transfers to or from a conventionally designated zero account or address. While functional, this lacked explicit block types for supply changes and didn't standardize fields for context (like a reason or the authorizer of the supply change) or facilitate easy querying of *only* mint/burn events.
> 
> - **ICRC-122**: This standard defines explicit `122mint` and `122burn` block types. The `tx` field within each block includes the `caller` principal who initiated the supply change, the essential data for the operation (account, amount), and an optional `reason` field for human-readable context. This allows for clearer auditing, validation, and direct querying of supply modification events and their authorizers recorded on the ledger.
> 

Author: ```suggestion
The `122mint` and `122burn` blocks introduced in ICRC-122 offer a more explicit and structured method for tracking token supply changes compared to how these operations are handled under the ICRC-1 standard. It is important to note that a ledger implementing both ICRC-1 and ICRC-122 may feature blocks resulting from both mechanisms, reflecting the distinct operations invoked.

- **ICRC-1 Mint/Burn Mechanism**:
    - **Operation & Block Type**: ICRC-1 defines a `minting_account`. Minting is achieved by an `icrc1_transfer` from this `minting_account` to the recipient, increasing total supply. Burning is an `icrc1_transfer` to the `minting_account`, decreasing total supply. ICRC-3 compliant ledgers supporting ICRC-1 mint and burn operations typically record these events using specific block types, often designated as `1mint` and `1burn` respectively (or similar, such as `icrc1_mint` and `icrc1_burn`). These blocks, while distinct from standard transfers, primarily confirm the transfer details associated with the minting or burning action as defined by ICRC-1.
    - **Authorization**: Authorization for minting is tied to the control of the `minting_account`. Only the principal that can initiate `icrc1_transfer` calls from this specific account can mint tokens. This centralizes minting authority to the controller of the minting account.
    - **Contextual Data**: The ICRC-1 `1mint` and `1burn` block structures, as typically derived from the underlying `icrc1_transfer` operation, do not have dedicated fields for a structured `reason` for the supply change beyond what might be placed in the generic `memo` field of the transfer. They also do not explicitly identify an ultimate `caller` or authorizer of the mint/burn operation beyond the `from` field of the transfer (which would be the minting account itself).

- **ICRC-122 Mint/Burn Blocks**:
    - **Explicit Block Types & Dedicated Methods**: This standard defines distinct block types (`122mint`, `122burn`) which result from calls to new, dedicated ledger methods (e.g., `icrc122_mint`, `icrc122_burn`). These are not special cases of `icrc1_transfer`.
    - **Flexible Authorization & Direct Caller Identification**: Authorization to call the `icrc122_mint` or `icrc122_burn` endpoints can be managed by the ledger implementation's governance logic, potentially allowing multiple, distinct whitelisted principals to initiate minting or burning. The `caller` field within the `tx` map of each `122mint` or `122burn` block directly records the principal that invoked this specific supply change operation.
    - **Enhanced Context**: The `tx` map includes an optional `reason` field, providing a standardized way to include human-readable context specifically for the supply change.
    - **Auditability**: The combination of dedicated block types, distinct methods, direct recording of the `caller`, and the optional `reason` field significantly enhances the auditability and transparency of token supply modifications compared to the ICRC-1 mechanism.

```

## Block Types & Schema

Each block introduced by this standard MUST include a `tx` field containing a map that encodes the token mint or burn transaction submitted to the ledger that caused this block to be created, similarly to how transaction blocks defined in ICRC-1 and ICRC-2 include the submitted transaction.

This enables canister clients, indexers, and auditors to reconstruct the exact instruction that led to the block being recorded to the ledger.

Each `122burn` or `122mint` block consists of the following top-level fields:

| Field    | Type (ICRC-3 `Value`) | Required | Description |
|----------|------------------------|----------|-------------|
| `btype`  | `Text`                 | Yes      | MUST be one of: `"122burn"` or `"122mint"`. |

> | `ts`     | `Nat`                  | Yes      | Timestamp in nanoseconds when the block was added to the ledger. |

Reviewer 2: I can't imagine there's any potential for confusion here, but should it be made explicit that this is Unix time/nanoseconds since the Unix epoch? (The same comment applies to the other standard drafts (ICRC-123, ICRC-124) as well.)

Author: Done here

> | `ts`     | `Nat`                  | Yes      | Timestamp in nanoseconds when the block was added to the ledger. |

Author: ```suggestion
| `ts`     | `Nat`                  | Yes      | Timestamp (in nanoseconds since the Unix epoch) when the block was added to the ledger. |
```

## Common Elements
This standard follows the conventions set by ICRC-3, inheriting key structural components.
- **Accounts** are represented using the ICRC-3 `Value` type, specifically as a `variant { Array = vec { V1 [, V2] } }` where `V1` is `variant { Blob = <owner_principal> }` representing the account owner, and `V2` is `variant { Blob = <subaccount> }` representing the subaccount. If no subaccount is specified, the `Array` MUST contain only one element (`V1`, the owner's principal).
- **Principals** (such as the `caller`) are represented using the ICRC-3 `Value` type as `variant { Blob = <principal_bytes> }`.
- Each block includes `phash`, a `Blob` representing the hash of the parent block, and `ts`, a `Nat` representing the timestamp of the block.

## Block Types & Schema

> Each block introduced by this standard MUST include a `tx` field containing a map. This map encodes information about the token mint or burn transaction, including the `caller` principal, the data required for balance calculation, and basic context.

Author: We have left authorisation out of this spec.

**Important Note on Transaction Recoverability:** The `tx` field defined below now includes the `caller` principal. However, for full auditability and transparency in complex scenarios, ledger implementations compliant with ICRC-122 **MUST** ensure that any other details of the original transaction invocation not captured in `tx` can be recovered independently. This could include, but is not limited to, the full arguments passed to the ledger method (if more complex than the data in `tx`), or any intermediary calls if the operation was part of a multi-step process. Mechanisms for recovering such extended data (e.g., via archive queries or specific lookup methods) remain implementation-dependent.

Each `122burn` or `122mint` block consists of the following top-level fields:

| Field    | Type (ICRC-3 `Value`) | Required | Description |
|----------|------------------------|----------|-------------|
| `btype`  | `Text`                 | Yes      | MUST be one of: `"122burn"` or `"122mint"`. |

> | `ts`     | `Nat`                  | Yes      | Timestamp in nanoseconds when the block was added to the ledger. |

Author: ```suggestion
| `ts`     | `Nat`                  | Yes      | Timestamp (in nanoseconds since the Unix epoch) when the block was added to the ledger. |
```

| `phash`  | `Blob`                 | Yes      | Hash of the parent block. |
| `tx`     | `Map(Text, Value)`     | Yes      | Encodes information about the token mint or burn transaction, including the caller. See schemas below. |

### `tx` Field Schemas

#### For `122burn`

| Field        | Type (ICRC-3 `Value`)                                        | Required | Description |
|--------------|--------------------------------------------------------------|----------|-------------|

> | `caller`     | `Value` (Must be `variant { Blob = <principal_bytes> }`)     | Yes      | The principal that invoked the ledger method (e.g., `icrc122_burn`) causing this block to be created. |

Author: The specific methods are left to implementations, but we do plan to have a spec that gathers the methods needed by RWA ledgers

| `from`       | `Value` (Must be `variant { Array = vec { V1 [, V2] } }`)¹ | Yes      | The account from which tokens are burned. |

> | `amount`     | `Nat`                                                        | Yes      | The number of tokens to burn. This value **MUST be greater than 0**. |

Author: Good point

This structure ensures that key information about the operation and its initiator is directly available in the `tx` field, enhancing transparency. For further details beyond what's in the block, such as the full original arguments passed to the ledger method (if applicable), implementations should provide separate recovery mechanisms as highlighted in the "Important Note on Transaction Recoverability."

## Compliance Reporting

Ledgers implementing this standard MUST return the following response to `icrc3_supported_block_types` with a URL pointing to the standard defining each block type:

```candid
vec {
    variant { Record = vec {
        record { "btype"; variant { Text = "122burn" }};
        record { "url"; variant { Text = "[https://github.com/dfinity/ICRC/blob/main/ICRCs/ICRC-122.md](https://github.com/dfinity/ICRC/blob/main/ICRCs/ICRC-122.md)" }}; // Placeholder URL
    }};
    variant { Record = vec {
        record { "btype"; variant { Text = "122mint" }};
        record { "url"; variant { Text = "[https://github.com/dfinity/ICRC/blob/main/ICRCs/ICRC-122.md](https://github.com/dfinity/ICRC/blob/main/ICRCs/ICRC-122.md)" }}; // Placeholder URL
    }};
}

```

> ## Example Blocks

Author: thanks!
